### ISSUE TITLE
RFC: lifting FLTK 1.4.x C++ Restrictions

### ISSUE THREAD

**Issue by MatthiasWM** [fltk/fltk#789]
Dear core developers and friends. According to our [Configuration Management Plan](https://www.fltk.org/cmp.php#CS_FLTK_1_1_X_RESTRICTIONS), FLTK 1.4 must not use:
* Exceptions
* Namespaces
* Standard C++ headers and library
* Templates
* dynamic_cast

And while I know the history and reasoning for those restrictions, I am very worried that we maneuver ourselves into a trap. I want to ask you all for consensus on allowing Standard C++ headers and library and dynamic_cast for 1.4 .

I am partially at fault at creating large chunks of code and duplicating C++ std functionality. `Fl_String` clones a subset of the  `std::string` API to the dot. As my excuse, writing FLUID without a String class has become extremely difficult. But there is also `Fl_Int_Vector` and almost 40 instances of `realloc` which, let's be honest, all just cover the lack of `std::vector`. Heck, we even have hardcoded array sizes in Fl_Help_View, Fl_Help_Dialog, and even Fl_Text_Buffer.

And we are *not* making FLTK small or light that way, we actually make it bigger. There is probably not a single C++ app out there that doesn't use std::string. So if we use it too, there is no additional code generated. But if we code around it all the time, we actually enlarge our binaries and our source code.

We are at a critical junction now. If we finalize 1.4, `Fl_String` will be out there in the wild, and we can never really take it back. So I suggest that we replace `Fl_String` with `std::string` *now* and *everywhere* in FLTK and FLUID. The same for `Fl_Int_Vector`.

I suggest that we allow `dynamic_cast<>` and runtime time information. RTTI is available on every compiler, and since it is just a sub feature of virtual calls, it does not add much to our library, but it keeps me from writing `as_window()` and similar silly workarounds. Even embedded systems like Arduino allow RTTI these days.

Don't get me wrong. I don't want to rewrite FLTK! I don't want to go in now and change the API to `std::string` for every `const char*`, but I really do want to replace code where we worked around issues that are easy to solve in modern C++.

I would still disallow exception. It can't be the job of a GUI library to generate potential application crashes with uncaught exceptions.

I would also discourage the excessive creation of new template classes. I am thinking about the authors of various language wrappers for FLTK (Rust, C, Python, Lua are active AFAIK) who surely can deal with a vector of strings, but would have a hard time with some esoteric template constructs that I have seem floating around in other libraries.

**Comment by MatthiasWM** [comment]
To sum it up, please vote on these issues:

1: allow specific std library classes and templates, notable `std::string` and `std::vector`
1a: immediately replace Fl_String with std::string and Fl_Int_Vector with std::vector<int>
2: allow `dynamic_cast<>` in FLUID and other apps within FLTK
3: allow all std library classes and templates for a specific C++ version
3a: require C++98, giving us the basics
3b: require C++11, adding `std::function` and `std::unique_ptr`, constexpr, nullptr, lambdas, enum classes, delegating constructors, range loops, non-static data member initializers, etc.
3c: require C++14, adding [[deprecated]]
3d: require C++17, adding `std::optional` and `std::filesystem` and a bunch more
3e: require C++20, 23, etc.?

I myself suggest requiring at least C++11. I'd prefer C++17. I am looking forward to your opinions!

**Comment by erco77** [comment]
I'd say if we do any of this, wait till 1.5 to do any retooling, as it'd be a big undertaking.

Since the goal here is to not suddenly prevent people from building FLTK on target platforms that lack the latest compiler features and tread carefully, perhaps after we decide which new syntax we want to support, we stick a few examples in some non-critical part of the 1.4.x code, e.g. a test program.

This way if someone encounters a problem building, it'll fail on building only that test program. They'll report it, but still have a built library. And we'll know who it tripped up and why (what compiler/platform they're on), and we can make decisions based on the results. We can put an #ifdef NO_TEST guard around it all, so we can recommend a build flag to help them rebuild without errors.

I'd suggest we include such test code in 1.4.x builds, and wait for *many* maintenance releases to go by (5 or 6 or so) before we start retooling code to allow enough sample environments to be tested.

**Comment by MoAlyousef** [comment]
Hi @MatthiasWM 

cfltk, the C bindings, use C++17 internally already. Things like auto, templates, range for loops, `using` instead of `typedef`, decltype, `constexpr if`, `static inline` etc are all life improvements. 

Using std classes in the public api, would be ok since these are not exposed by the C bindings, however, for fellow C++ devs, it would limit them to using the same std library FLTK was built against.
i.e Fl_Widget::label(const std::string &l) would cause linking errors if FLTK was built with g++ (using libstd++), and other devs tried to pass a libcxx std::string.

**Comment by MatthiasWM** [comment]
@MoAlyousef @erco77 Thanks for you answers. As I wrote [in the other issue report](https://github.com/fltk/fltk/issues/789), I am not suggesting to mess much with the existing API (although I have made Fl_String available in fl_filename_*). I just really want to avoid that users start to write code with Fl_String and Fl_Int_Vector, and we will never get rid of that code duplication.

@erco77 I know how reluctant FLTK is to change, but std::string has been around for over 25 years now, and I just can't imagine a single project that would not support it.

> would cause linking errors if FLTK was built with g++ (using libstd++), and other devs tried to pass a libcxx std::string.

@MoAlyousef How do other libraries deal with that issue? I mean we are hardly the only library in the C++ world that considers using std::string, right?

**Comment by clbr** [comment]
I am for the std:: classes but against rtti and exceptions. The size impact of -fno-rtti -fno-exceptions is quite measurable.

**Comment by MatthiasWM** [comment]
4: if we do not allow std::string, should we remove any API that uses Fl_String? Maybe make Fl_String entirely private? Maybe make Fl_String only part of FLUID, but not FLTK?

I am just afraid that once we officially release Fl_String, we can never take it back.

**Comment by ManoloFLTK** [comment]
I support to delay the introduction of the std library into FLTK until a future major version.

I'm concerned by @MoAlyousef 's point of potential conflicts between different standard library implementations used by FLTK and by FLTK-using apps. Is the only way out this one :
>  A compromise would be to limit std types to internal use, or to wrap them using an opaque type.


Introducing class `Fl_String` in 1.4 knowing already that it's condemned to death makes no sense to me. A solution would be to move file `Fl_String.cxx` to `fluid/` and use class `Fl_String` only in the fluid code. In the future FLTK 4.0, fluid would just use `std::string` or would `typedef std::string Fl_String`.

Functions `fl_input_str()` and `fl_password_str()` could be renamed something like `fl_XXX_persistent()`, return a `malloc`'ed `const char *`value and be documented to require a call to `free()` after use.

**Comment by ManoloFLTK** [comment]
@Albrecht-S I understood all of that. My main point remains that creating class `Fl_String` while knowing it will disappear as soon as possible seems to me a bad idea.

**Comment by Albrecht-S** [comment]
@MatthiasWM wrote:
> I just tried that with master, and it actually worked without a hitch! In Fl_Preferences.H, one needs to replace the forward class declaration with #include <Fl_String.H>, but that's all.

I can't confirm this entirely, but the small differences between `Fl_String` and `std::string` should all be fixable. I got a few warnings and even errors when compiling, particularly but not only in `test/unittest_core.cxx`. My compilation was done with C++11, later C++ standards may be different.

Another test with C++98 revealed that C++98 doesn't provide `std::string::pop_back()` whereas C++11 has it. This wouldn't be an issue if we required at least C++11 in FLTK 4.0. We might want to remove this if we required only C++98.

There are also some `Fl_String` specific extensions, e.g. `Fl_String::strlen()` and `Fl_String::shrink_to_fit()` which are not implemented in `std::string`. We might want to remove these methods.

Other than that, I agree that there are only minor issues which we may be able to fix so `Fl_String` is really compatible with `std::string`, at least since `C++11` and that can be used in FLTK 4.

FWIW: I'm attaching a patch that I used on my system (Debian Bookworm, g++ 12.2.0) to make our sources compile w/o warnings and errors. This patch doesn't really fix anything, it only works around errors and warnings.

[Fl_String.diff](https://github.com/fltk/fltk/files/12847801/Fl_String.diff.txt)

Note that this is not an argument whether or not to keep `Fl_String` (which I recommend, BTW) but only a small test for compatibility between `Fl_String` and its potential later replacement `std::string`.

**Comment by MatthiasWM** [comment]
Ok, so we have in FLTK:

 * Fl_Preferences::set and get
 * fl_filename_* multiple, but already available with const char *
 * fl_input_str 2x
 * fl_password_str 2x
 * private class inside Fl_Help_View

On the FLUID side, I can't turn things back, but I can move `Fl_String`/`Fl_Preferences`/`fl_filename_*` over in the fluid directory and be happy.

As for the four dialogs, these were Albrecht's wishes. They are a bit clumsy because on Cancel, they can't return `NULL`. With C++, we would have `std::optional<std::string>` which is similar to what Rust generally does if a value or an error could be returned, or we could return a tuple of a result code and a string (but that's even newer C++). I see how they are convenient, but they are also just duplicates of functions return `const char *`.

**Comment by Albrecht-S** [comment]
> On the FLUID side, I can't turn things back, but I can move Fl_String/Fl_Preferences/fl_filename_* over in the fluid directory and be happy

I agree that you can move `Fl_String` over to fluid (or maybe to `src`), but I don't think that `Fl_Preferences` and `fl_filename_*` can just be moved. Admitted, you can move those parts of fl_filename* with `Fl_String` to fluid but `Fl_Preferences` is a class. Well, you could derive your own subclass if we removed the `Fl_String` methods from the public `Fl_Preferences` class. Not nice but doable. In FLTK 4 we could then add the std::string variants to the public Fl_Preferences API.

> As for the four dialogs, these were Albrecht's wishes. They are a bit clumsy because on Cancel, they can't return NULL.

In the original implementation they could "return NULL" somehow (`c_str()` could be NULL). However, we dropped this for compatibility with `std::string` and the new implementation with an additional error argument is IMHO OK (so far). But as I wrote above, if the majority of devs want to remove `Fl_String` from the public interface (make it private), then so be it. We can rewrite `fl_input_str()` and `fl_password_str()` to return `(const) char *` with the requirement to free the pointer after return (by user code). OK.

The private class inside `Fl_Help_View` shouldn't be an issue, it derives from `Fl_String` which would still be internally available. Nothing of this is public.

We also have the non-standard but public `Fl_Int_Vector` class. I made a proof of concept:
- move the header to the `src` folder and
- rewrite `Fl_Table` that uses it and
- include "Fl_Int_Vector.H" in other classes from the `src` folder.

This would hide `Fl_Int_Vector` completely from the public interface. Seems to work but needs some cleanup and testing.

> as @erco77 suggested, since FLUID ~as~ is not the library but an app, I could simply use std::string and ...

Nope, fluid must be compilable by users, even if cross-compiling or whatever. Please keep this in mind.

> ... "sneak a little bit of the std library into FLTK unit someone complains"

??? I don't understand this part.

**Comment by ManoloFLTK** [comment]
> Ok, so Fl_Int_Vector goes into src, Fl_String goes into fluid, fl_filename_* with Fl_String results also in fluid, and the dialogs get an additional version with matcher, but the Fl_String versions are removed until we allow std::string?

This sounds very good to me.

**Comment by Albrecht-S** [comment]
> >  ... I don't think that Fl_Preferences and fl_filename_* can just be moved.

> Well, they both would expose Fl_String in a public header. Let's just move those two and move them back when we have std::string available.

@MatthiasWM Sure, the current versions expose `Fl_String`, but ... either I'm misunderstanding you or you are misunderstanding me (?). You can't **_move_** `Fl_Preferences.H` (!) to a private location because `Fl_Preferences` **is and has always been a public class**. If you want to use the methods that use `Fl_String` in fluid, my point is that you need to subclass and extend `Fl_Preferences` in the `fluid` folder. Or am I missing anything?

Other than that, yes, that's OK with me.

`fl_filename_*` is different though. I think you can move only those methods that use Fl_String to an internal location (fluid), that's not a problem. Or something like that.

> the dialogs get an additional version with matcher, but the Fl_String versions are removed until we allow std::string?

Yes, I agree, if "matcher" means "maxchar" ;-) . This "maxchar" addition was requested by an STR IIRC.

**Comment by Albrecht-S** [comment]
Addition: `Fl_String.H` should be moved to the `src` folder because it is needed in `Fl_Help_View` as well.

**Comment by RokerHRO** [comment]
I'd favor all features of the language and the stdlib that have no runtime costs (neither CPU nor RAM) compared to self-written alternatives of C++17 at least:
* `std::string` and even more `std::string_view`
* `std::any`
* `std::vector<T>, `std::map` (maybe more containers if necessary & sensible)
* `std::optional<T>`, `std::variant<…>`
* `std::clamp()`
* stdandard algorithms like `std::sort()`, `std::transform()`, `std::for_each()`, `std::fill()` etc.

https://en.cppreference.com/w/cpp/17

Other features especially of the runtime library might not be present at all platforms FLTK supports (needs further investigation/documentation):
* std::filesystem
* std::thread
* std::mutex and locks

**Comment by RokerHRO** [comment]
> I'd say if we do any of this, wait till 1.5 to do any retooling, as it'd be a big undertaking.
> 
> Since the goal here is to avoid suddenly preventing users from building FLTK on target platforms that lack the latest compiler features, we'll want to tread carefully.

ACK. So which platforms are FLTK willing to support? Which platforms are currently in use and/or "relevant" for FLTK development? Are there any reliable data?

> Perhaps after we decide which new syntax we want to support, we stick a few examples in some non-critical part of the 1.4.x code, e.g. a test program, and see if anyone reports build failures for that test program.

Good idea! 👍 

> I'd suggest we include such test code in 1.4.x builds, and wait for _many_ maintenance releases to go by (5 or 6 or so) to get a large sample of production environments tested. Then we can reconsider, if need be, based on whichever features trip up the most.

I'd favor to do the modernization right now, with the new 1.4 release. 1.3.x is still in use on many old, but living platforms, so the 1.3.x branch can still use the "C-like" API.

But specifying: "For FLTK 1.4.x you need a C++11/14/17 compiler (maybe lacking feature X and Y of the runtime library)" is now the right moment, I'd say.

Later releases of 1.4.x might use gently more and more modern C++ features, when they are needed,  sensible or desired.

**Comment by MatthiasWM** [comment]
Hmm, so to make a good decision, how long until we can even think about releasing 1.4.0 ? If the release is (somewhat) imminent, I suggest that we do the changes to Fl_String and Fl_Int_Vector to no pollute our API, and then try to get 1.4.0 out of the door ASAP.

I the suggest a split approach and allow fixes in 1.4.x, introduce std::* in 1.5.0/master, and have a separate discussion about 4.0. That way, we can release a safe 1.4.0, still get modern C++ into master within bearable time span, and have as much time as we like to think about a 4.0 .

@Albrecht-S and @all : do we have a singular comprehensive and decisive list for what is needed to release 1.4.0? I am willing to drop all my other little improvement plans on FLUID for a quick 1.4.0 release.

**Comment by RokerHRO** [comment]
> > On which platform? Can you give us up-to-date benchmarks to have a robust base for discussion, please?
> 
> On Linux. x86_64, fltk 1.3.5, gcc 10.2, a -fno-rtti -fno-exceptions build is about 3% smaller in both static and shared configs.

Okay, thank you for the answer! :-)

Whether 3% are worth to waive important C++ features and software design patterns or not, is worth to discuss about.

But I thought about embedded platforms where no-RTTI have a much more serious impact, or in other words: "RTTI is inacceptable".

**Comment by RokerHRO** [comment]
> @clbr @RokerHRO The impact of RTTI is one additional pointer in the vtable per available and linked type, and some implementations still have this pointer even if you compile with no-rtti. https://stackoverflow.com/questions/579887/how-expensive-is-rtti .

"Blast from the past". ;-)
The thread is 14 years old. And the benchmark code given is misleading, because it allows out-of-loop optimizations in one approach that is not possible in the other, as mentioned in the answers.

So I'd like to mention one of the most recent answers: https://stackoverflow.com/a/76378353 and I agree with this guy.

**Comment by clbr** [comment]
I'm in general opposed to bloat, that's precisely what leads to "modern" things like an editor that takes 5 GB of RAM. Exceptions are in my opinion a bad pattern, and there was agreement before that a library shouldn't use them. RTTI can be useful in places, but the use in FLTK would be in a handful of places, which in no way justifies the bloat of several percent.

I personally don't use FLTK in "deep embedded", but in TinyCore we do operate on old computers.

**Comment by clbr** [comment]
Re the earlier Arduino comment, what they do should not be taken as a good approach or a validation. After all they use python. When I was working on a past project on an esp mcu, it was a common occurrence in the forums to find "why is it so slow" "why can't I do this that this other project is doing" "why does it take this long to answer an interrupt" and the answer was because they were using arduino, which was greatly slower vs C or even asm.

So in other words, Arduino is the easy beginner way, but greatly not performant nor size-conscious. Users hitting various limits with it then proceeded to native C, or dropped the project (rarely buying a faster board would have given enough speedup).

**Comment by Albrecht-S** [comment]
@MatthiasWM wrote:
> to make a good decision, how long until we can even think about releasing 1.4.0 ? If the release is (somewhat) imminent, I suggest that we do the changes to Fl_String and Fl_Int_Vector to no pollute our API, and then try to get 1.4.0 out of the door ASAP.

I'd say release 1.4.0 is imminent, i.e. we should strive to get it out in one or two months, at least at the end of the year. I'll try my best to get my open development jobs done until end of November but I can't guarantee to get everything on my todo list done until then.

My personal deadline is the end of this year, and if there are still open bugs I think we should release 1.4.0 anyway. It's such an improvement compared to 1.3 that it's now really time to get it out.

PS: I'll be busy for the rest of today and tomorrow, but I'll reply more to the latest comments when I can afford the time.

**Comment by Albrecht-S** [comment]
> do we have a singular comprehensive and decisive list for what is needed to release 1.4.0?

Not currently an up-to-date list, but there is [our milestone "Release FLTK 1.4.0"](https://github.com/fltk/fltk/milestone/1) which should give some hints and of course all currently open GitHub issues that are important enough to require a fix.

**Comment by Albrecht-S** [comment]
@RokerHRO wrote:
> I'd favor to do the modernization right now, with the new 1.4 release. 1.3.x is still in use on many old, but living platforms, so the 1.3.x branch can still use the "C-like" API.

I believe we have **_consensus_** to do this **_soon but not yet in 1.4.x_**.

The main reason is that we don't want to preclude users of embedded systems (cross-compilers) and other very old systems from building FLTK 1.4. Our policy has always been not to break user code and to be backwards compatible (in this case to FLTK 1.3). The lack of real user data, i.e. _"who uses which old compiler"_ and _"which C++ features can't be used by (some of) our users"_ forces us to keep the old policy for 1.4. I plan to open a "poll" on the FLTK website soon to ask users about their current development restrictions so we get a better picture for the future.

We developed FLTK 1.4 for many years and we can release it very soon as it is. Some bug fixes and new features (e.g. `Fl_Flex`, `Fl_Grid`, much improved `fluid`, a new app `fltk-options`, and last but not least `Wayland` support) will be added and all our current (1.3.x and 1.4/git) users and testers will be able to use these features. Since 1.4 is 100% backwards compatible to 1.3 and does NOT require new compiler features we can safely abandon FLTK 1.3 development because ALL users can move forward to 1.4 w/o problems.

After the release of 1.4.0 we can (IMHO) very quickly switch to more modern C++ features in the next minor (1.5.0) or maybe even major (4.0.0) branch/version. This will be announced in the release of FLTK 1.4 and users that can't follow the development will have to stay with 1.4.x which is much better than 1.3.x.

We should however wait some time to get enough feedback from users of 1.4.0 and fix issues in one or more maintenance releases `1.4.x` until we are confident that `1.4.x` doesn't need more fixes in the short term.

Once we decide that FLTK 1.4.x is stable enough we can start the new branch. Waiting some time after the first release of such a big update is important to avoid maintaining two branches simultaneously. How long this waiting period will be I can't tell now, but I hope this will be no more than a few months. This depends on the number of bug reports we receive. My plan is to "release often" during that time frame. Please keep also in mind that we will publish one or more release candidates before 1.4.0 will be released.

My further intention for the "next gen FLTK" is to open it for selected new C++ features as discussed above. As said before, we really need `std::string`, `std::vector`, and some other features like `auto` variables. This could all be done internally, i.e. w/o changing the API at all but we could also add some new features like returning `std::string` and/or allowing `std::string` as input parameters. The main reason to open FLTK for modern C++ is to get rid of such things like `Fl_Int_Vector` and `Fl_String` as discussed above and to **_simplify code maintenance_**.

BTW, I remember that we got some PR's where users suggested to use `std::vector` internally which we couldn't accept just because of our current policy. One of these PR's has been implemented with `Fl_Int_Vector`, for instance. Fortunately this is not in the public API.

That said, we could basically release 1.4.0 soon and "immediately" start the new branch w/o changing a single statement - just document the new requirements (for instance C++11 or whatever) and slowly start rewriting some critical internal code, i.e. replace `Fl_Int_Vector` with `std::vector` and `Fl_String` with `std::string`, and that's it. We could even (theoretically) release this as 1.5.0 within days or weeks after 1.4.0 (but see above for the "feedback" period).

@MatthiasWM wrote:
> I ~the~ suggest a split approach and allow fixes in 1.4.x, introduce std::* in 1.5.0/master, and have a separate discussion about 4.0. That way, we can release a safe 1.4.0, still get modern C++ into master within bearable time span, and have as much time as we like to think about a 4.0 .

I agree with that statement (this **could** be done) but we should discuss the further development separately because this issue is about 1.4.x after all.

(more in my next comment)

**Comment by Albrecht-S** [comment]
Summary of changes as discussed above which (I hope) we have found consensus about.

**_To do before the release of 1.4.0:_**

- [ ] Make `Fl_Int_Vector.H` private by moving it to the `src` folder (Albrecht)
- [ ] Make `Fl_String.H` private by moving it to the `src` folder (Albrecht)
- [ ] Rewrite `fl_input_str()` and `fl_password_str()` to return `const char *` (Albrecht)
- [ ] Refactor `Fl_Table` to remove `Fl_Int_Vector` from its public header `FL/Fl_Table.H` (Albrecht)
- [ ] Move `fl_filename*` functions that use `Fl_String` to `fluid` (or `src` ?) (Matthias)
- [ ] Move `Fl_Preference` methods that use `Fl_String` to `fluid` (or `src` ?) (Matthias)

My plan is to commit and push points 1 and 2 (simply moving the headers and changing the include statements) in a first step so different devs can work on the other points. I would also like to take points 3 and 4 because I was the original author of the `fl_*_str()` methods and I worked already on a proof of concept for `Fl_Table`.

@MatthiasWM I would also suggest that you take the other two points, OK?

**_To do in the "next gen branch" (after the release of 1.4.0):_**

- Add methods that return std::string to fl_input* and fl_password* etc.
- Restore `fl_filename*` functions using `std::string` rather than `Fl_String`
- Restore `Fl_Preference` methods using `std::string` rather than `Fl_String`

More details of the "next gen branch" would be OT for this issue but many thanks to all who contributed their thoughts. We should discuss the further development after 1.4.x separately either in `fltk.coredev` or in another GitHub issue.

If nobody objects (please reply below whether you agree or disagree) I will start quickly with points 1 and 2 and after that on 3 and 4.

**Comment by ManoloFLTK** [comment]
Very good.

I add that class `Fl_Int_vector` is also used by file `src/drivers/Wayland/Fl_Wayland_Screen_Driver.cxx`. It will be enough to change the corresponding `#include` clause to `#include "../../Fl_Int_Vector.H"`.

**Comment by Albrecht-S** [comment]
> I add that class Fl_Int_vector is also used by file src/drivers/Wayland/Fl_Wayland_Screen_Driver.cxx.

Yup, thanks for the hint. That's already done in my POC. I'm generally searching for include files like this:
```
$ git grep -n 'include.*Fl_Int_Vector\.H'
src/Fl_Help_View.cxx:53:#include "Fl_Int_Vector.H"
src/Fl_Int_Vector.H:36:  #include <FL/Fl_Int_Vector.H>  // documentation not yet updated
src/Fl_Int_Vector.cxx:18:#include "Fl_Int_Vector.H"
src/Fl_Table.cxx:23:#include "Fl_Int_Vector.H"
src/drivers/Wayland/Fl_Wayland_Screen_Driver.cxx:29:#include "../../Fl_Int_Vector.H"
```
which should find all relevant occurrences.

**Comment by Albrecht-S** [comment]
Points 1, 2, and 4 done in my local branch (still WIP, but seems to work well so far).

I did a small benchmark with `Fl_Table`: allocating 1 million rows and 1 million cols with our internal `Fl_Int_Vector` and with `std::vector<int>` using preprocessor macro `USE_STD` to be able to use both alternatively.

Values of a single test (not representative):
```
Fl_Table::rows(1000000) from 1, USE_STD = 0   <<< Fl_Int_Vector
Fl_Table::rows(1000000) - done in   8.610 ms
Fl_Table::cols(1000000) from 1, USE_STD = 0
Fl_Table::cols(1000000) - done in   8.437 ms
...
Fl_Table::rows(1000000) from 1, USE_STD = 1   <<< std::vector<int>
Fl_Table::rows(1000000) - done in   8.202 ms
Fl_Table::cols(1000000) from 1, USE_STD = 1
Fl_Table::cols(1000000) - done in   5.783 ms
```

Values are similar but `std::vector` seems to "win" (faster).

BTW: Measurement was done with `Fl::now()` and `Fl::seconds_since().` Great to have this feature, it's really easy to use. Thanks to @MatthiasWM.

**Comment by Albrecht-S** [comment]
Another measurement, this time with 10 Mio rows and cols, starting at 500 each, including CPU time.
```
Fl_Table::rows(10000000) from 500, USE_STD = 0   <<< Fl_Int_Vector
Fl_Table::rows(10000000) - done in  70.631 ms
Fl_Table::cols(10000000) from 500, USE_STD = 0
Fl_Table::cols(10000000) - done in  70.868 ms

user	0m0.788s
sys	0m0.165s

Fl_Table::rows(10000000) from 500, USE_STD = 1   <<< std::vector<int>
Fl_Table::rows(10000000) - done in  43.329 ms
Fl_Table::cols(10000000) from 500, USE_STD = 1
Fl_Table::cols(10000000) - done in  33.989 ms

user	0m0.705s
sys	0m0.115s
```
Here std::vector seems to be much better!

**Comment by Albrecht-S** [comment]
@MatthiasWM and other devs: after my recent tests I got a new idea for FLTK 1.4: we could add a build option that allows to use `std::vector` and `std::string` whose default would be OFF.

- If the option is **_ON_** we can **replace** all functions that use `Fl_String` with arguments and return types `std::string`.
- If the option is **_OFF_** we work as described above. Functions using `Fl_String` will only be available in Fluid.

Instead of **removing** the `Fl_String` methods and functions we can put them in `#if HAVE_STD_STRING ... #else ... #endif` conditionals or something like that.

I did something similar for my test with `Fl_Table` and `Fl_Int_Vector`. If the option is **_ON_** `Fl_Table` uses `std::vector<int>`, otherwise it uses `Fl_Int_Vector`.

This way we can have basic `std::*` support in FLTK 1.4.x (opt-in) and prepare for 1.5 or 4.0 when we just need to disable the conditional code and switch to `std::*` completely. Similar to what we did with `FL_ABI_VERSION` in 1.3.x.

**Comment by RokerHRO** [comment]
@Albrecht-S : Great benchmark! Because some developers mentioned different code size: How does the code size differ in your example (if the rest of the program does not use `std::vector` nor `std::string`, of course)?

**Comment by clbr** [comment]
> @MatthiasWM and other devs: after my recent tests I got a new idea for FLTK 1.4: we could add a build option that allows to use `std::vector` and `std::string` whose default would be OFF.

I don't believe such an option would be useful to users. Anyone with a
specific setup can test the new branch just as easily.

**Comment by ManoloFLTK** [comment]
> we could add a build option that allows to use std::vector and std::string whose default would be OFF.

This looks very complicated for an extremely small potential use since the difference would apply only inside fluid, not in the public FLTK API.

**Comment by Albrecht-S** [comment]
> How does the code size differ in your example (if the rest of the program does not use std::vector nor std::string, of course)?

The test program is `test/table.cxx` (unchanged) and the only modification is that `Fl_Table` uses either `Fl_Int_Vector` or `std::vector<int>` in two instances. Both objects are allocated with operator `new` in both versions. The `std::vector` version is slightly larger. My build was with debug option and the figures below show both versions before and after stripping.
```
                debug:             stripped:             
std::vector:    4975392 / 4.8M     1170520 / 1.2M
Fl_Int_Vector:  4917552 / 4.7M     1162264 / 1.2M
```
The difference of the stripped version is only 8256 bytes, I'd say this is a minimal overhead for
- safety (less error prone)
- better performance
- developer convenience (easier and shorter coding).

**Comment by Albrecht-S** [comment]
@clbr wrote:
> I don't believe such an option would be useful to users. Anyone with a specific setup can test the new branch just as easily.

We offer some really useful new methods that input or output `std::string` (currently `Fl_String` but these methods would otherwise be removed). As it turned out these methods were essential for the development of new features in fluid - other application developers may have similar issues and appreciate FLTK offering std::string even if it is only for a few methods.

Many users can't use inofficial branches and would therefore need an official released version. Such users could use the new build option and benefit from the new std::string using functions.

Since the effort to enable this option is minimal I believe strongly that we could and **should** do it.

**Comment by ManoloFLTK** [comment]
> So what we need to do is only add the #ifdef /#endif around these few methods and replace Fl_String with std::string (literally).

If I understand correctly, class `Fl_String` would not be present in the public API, neither activating nor deactivating the new proposed build option. In that case, it's a very good proposal.

**Comment by MatthiasWM** [comment]
> ```
>                 debug:             stripped:             
> std::vector:    4975392 / 4.8M     1170520 / 1.2M
> Fl_Int_Vector:  4917552 / 4.7M     1162264 / 1.2M
> ```

It's my understanding that duplicate code will be dropped by the linker, so if we link statically to an app that uses `std::vector<int>` elsewhere, the app will actually be *smaller* than with `Fl_Int_Vector`. Is that so?

**Comment by Albrecht-S** [comment]
> If I understand correctly, class Fl_String would not be present in the public API, neither activating nor deactivating the new proposed build option. In that case, it's a very good proposal.

Yes, that is correct.

1. If the new option is **ON** selected API's, particularly those that used `Fl_String`, will be enabled but use `std::string` **instead of** `Fl_String`.

2. If the new option is **OFF**
- those functions and methods that used `Fl_String` will be deleted from the public API and only be available internally, especially for `fluid`
- neither `Fl_String` nor `std::string` will be visible in the public API
- `std::string` will not be used anywhere in the FLTK code for backwards compatibility.

**Comment by Albrecht-S** [comment]
FYI: I'm working on the removal of Fl_String from the public interface. ...
More to follow when I have everything working.

**Comment by Albrecht-S** [comment]
As you may have noticed, I pushed a big commit (1209e9dcd7e1e97bedc747d06ba4eea837562158) that removes `Fl_String` and `Fl_Int_Vector` from the public interface (header files). Both classes are still available for internal usage.

I also added new CMake and configure options so we can selectively add some new functions and methods that use - at least - `std::string` in the public interface if and only if the option is active. Look for `FLTK_USE_STD` in the code if you're interested. Current usage is very restricted. There are only a few methods that previously used `Fl_String` which I could "convert" to use `std::string` if the option is ON.

Fluid still uses `Fl_String`, and so does `Fl_Help_View`. `Fl_Int_Vector` is also used in two widgets internally.

I believe that this commit is basically enough to close this issue and a milestone on our way to release 1.4.0. However, there might be some missing pieces or bugs, therefore I would like to keep it open for a while and wait for feedback. Thanks in advance.

**Comment by ManoloFLTK** [comment]
Well done.

I tried to replace
```
#include "../Fl_Int_Vector.H"
by
#include <vector>
typedef std::vector<int> Fl_Int_Vector;
```
in files `Fl_Help.cxx` and `Fl_Wayland_Screen_Driver.cxx`. In both cases, use of the `pop_back()` member function produces errors. We did well in not letting class `Fl_Int_Vector` go public.

**Comment by Albrecht-S** [comment]
> use of the pop_back() member function produces errors. We did well in not letting class Fl_Int_Vector go public.

Yes, I agree. As I found out (IIRC) std::vector::pop_back() was added in a later C++ standard than C++11, maybe C++14 or C++17. But it exists in these versions. Note that I replaced the code in question in Fl_Table in the case OPTION_USE_STD is enabled.

**Comment by MatthiasWM** [comment]
Thanks for all the work. I am glad we didn't put anything out there that we can't retract!
